### PR TITLE
#4508 writing tests for zookeeperEpics

### PR DESCRIPTION
**Pull request by saivirtue**
### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #4508)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [x] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-manager/client/src/api/__mocks__/inspectApi.ts [hunk 1]
```diff
@@ -49,10 +49,42 @@ export const entity = {
...
+      defaultValue: null,
+      recommendedValues: [],
+      group: 'core',
+      prefix: '',
+    },
+  ],
+  classInfos: [],
+};
+
 // simulate a promise request with delay 3s
```
Comment (konekoya) [inline ohara-manager/client/src/api/__mocks__/inspectApi.ts:77]: _ with a delay of 3s_ sounds better to me :)

File: ohara-manager/client/src/api/__mocks__/inspectApi.ts [hunk 2]
```diff
@@ -49,10 +49,42 @@ export const entity = {
   classInfos: [],
 };
 
+export const zookeeperInfoEntity = {
+  imageName: 'oharastream/zookeeper:0.9.0',
```
Comment (konekoya) [inline ohara-manager/client/src/api/__mocks__/inspectApi.ts:53]: This image name is different from the above? Is that intentional?

File: ohara-manager/client/src/store/epics/__tests__/startZookeeperEpic.test.js [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+    });
+
+    expectSubscriptions(action$.subscriptions).toBe(subs);
+
+    flush();
+  });
+});
+
+it('start zookeeper failed after reach retry limit', () => {
+  // mock a 20 times "failed started" results
```
Comment (konekoya) [inline ohara-manager/client/src/store/epics/__tests__/startZookeeperEpic.test.js:95]: result should be singular

File: ohara-manager/client/src/store/epics/__tests__/startZookeeperEpic.test.js [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+        payload: zookeeperEntity,
+      },
+    });
+    const output$ = startZookeeperEpic(action$);
+
+    expectObservable(output$).toBe(expected, {
+      a: {
+        type: actions.startZookeeper.REQUEST,
+        payload: {
+          zookeeperId: getId(zookeeperEntity),
```
Comment (konekoya) [inline ohara-manager/client/src/store/epics/__tests__/startZookeeperEpic.test.js:179]: Do you consider use a var to store the this? As it's used in many places :)

File: ohara-manager/client/src/store/epics/__tests__/stopZookeeperEpic.test.js [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+ */
+
+import { TestScheduler } from 'rxjs/testing';
+
+import stopZookeeperEpic from '../zookeeper/stopZookeeperEpic';
+import * as zookeeperApi from 'api/zookeeperApi';
+import { entity as zookeeperEntity } from 'api/__mocks__/zookeeperApi';
+import * as actions from 'store/actions';
+import { getId } from 'utils/object';
+import { of } from 'rxjs';
```
Comment (konekoya) [inline ohara-manager/client/src/store/epics/__tests__/stopZookeeperEpic.test.js:24]: Third lib...

Comment (konekoya) [reply ohara-manager/client/src/store/epics/__tests__/stopZookeeperEpic.test.js:24]: party

File: ohara-manager/client/src/store/epics/zookeeper/startZookeeperEpic.js [hunk 6]
```diff
@@ -36,20 +36,30 @@ import * as actions from 'store/actions';
...
+const startZookeeper$ = params => {
   const zookeeperId = getId(params);
   return of(
     defer(() => zookeeperApi.start(params)),
     defer(() => zookeeperApi.get(params)).pipe(
       map(res => {
-        if (!res.data?.state) throw res;
-        if (res.data.state !== SERVICE_STATE.RUNNING) throw res;
-        return res.data;
+        if (!res.data?.state || res.data.state !== SERVICE_STATE.RUNNING)
```
Comment (konekoya) [inline ohara-manager/client/src/store/epics/zookeeper/startZookeeperEpic.js:45]: shouldn't this be `!res.data?.state || res.data?.state !== SERVICE_STATE.RUNNING`?

File: ohara-manager/client/src/store/epics/zookeeper/stopZookeeperEpic.js [hunk 7]
```diff
@@ -17,37 +17,47 @@
...
-export const stopZookeeper$ = params => {
+const stopZookeeper$ = params => {
   const zookeeperId = getId(params);
   return of(
     defer(() => zookeeperApi.stop(params)),
     defer(() => zookeeperApi.get(params)).pipe(
       map(res => {
         if (res.data?.state) throw res;
-        return res.data;
+        else return res.data;
```
Comment (konekoya) [inline ohara-manager/client/src/store/epics/zookeeper/stopZookeeperEpic.js:45]: else can be omitted, but if you're being explicit, then ignore me... 😅

File: ohara-manager/client/src/store/epics/zookeeper/updateZookeeperEpic.js [hunk 8]
```diff
@@ -17,21 +17,15 @@
...
+import { defer, of } from 'rxjs';
+import { catchError, map, startWith, mergeMap } from 'rxjs/operators';
 
 import * as zookeeperApi from 'api/zookeeperApi';
 import * as actions from 'store/actions';
 import * as schema from 'store/schema';
 import { getId } from 'utils/object';
 
-export const updateZookeeper$ = values => {
+const updateZookeeper$ = values => {
```
Comment (konekoya) [inline ohara-manager/client/src/store/epics/zookeeper/updateZookeeperEpic.js:28]: Good catch :)

File: ohara-manager/client/src/api/__mocks__/inspectApi.ts (hunk 1 above)
Comment (saivirtue) [reply ohara-manager/client/src/api/__mocks__/inspectApi.ts:77]: Thanks for the correction :)
I will update the similar part of the other files.

File: ohara-manager/client/src/api/__mocks__/inspectApi.ts (hunk 2 above)
Comment (saivirtue) [reply ohara-manager/client/src/api/__mocks__/inspectApi.ts:53]: No. Just copy from our [doc site](https://ohara.readthedocs.io/en/latest/) 😅

File: ohara-manager/client/src/store/epics/__tests__/startZookeeperEpic.test.js (hunk 3 above)
Comment (saivirtue) [reply ohara-manager/client/src/store/epics/__tests__/startZookeeperEpic.test.js:95]: OK

File: ohara-manager/client/src/store/epics/__tests__/startZookeeperEpic.test.js (hunk 4 above)
Comment (saivirtue) [reply ohara-manager/client/src/store/epics/__tests__/startZookeeperEpic.test.js:179]: Hmm...good suggestion

File: ohara-manager/client/src/store/epics/__tests__/stopZookeeperEpic.test.js (hunk 5 above)
Comment (saivirtue) [reply ohara-manager/client/src/store/epics/__tests__/stopZookeeperEpic.test.js:24]: 😅

File: ohara-manager/client/src/api/__mocks__/inspectApi.ts (hunk 2 above)
Comment (konekoya) [reply ohara-manager/client/src/api/__mocks__/inspectApi.ts:53]: No problem :)

File: ohara-manager/client/src/store/epics/zookeeper/startZookeeperEpic.js (hunk 6 above)
Comment (saivirtue) [reply ohara-manager/client/src/store/epics/zookeeper/startZookeeperEpic.js:45]: If state is `undefined`, the former condition will be `false`. Hence the execution will never go to the later one :)

File: ohara-manager/client/src/store/epics/zookeeper/stopZookeeperEpic.js (hunk 7 above)
Comment (saivirtue) [reply ohara-manager/client/src/store/epics/zookeeper/stopZookeeperEpic.js:45]: 👍

File: ohara-manager/client/src/store/epics/zookeeper/startZookeeperEpic.js (hunk 6 above)
Comment (konekoya) [reply ohara-manager/client/src/store/epics/zookeeper/startZookeeperEpic.js:45]: 👍

**Pull request merged**